src/stcal/jump/jump.py

```diff
@@ -846,221 +738,92 @@ def near_edge(jump, low_threshold, high_threshold):
           Group dq array.
+
       readnoise_2d : float, 2D array
           Readnoise for all pixels.
-      nframes : int
-          The number frames that are averaged in the group.
-      snr_threshold : float
-          The signal-to-noise ratio threshold for detection of extended
-          emission.
-      min_shower_area : int
-          The minimum area for a group of pixels to be flagged as a shower.
-      inner : int
-          The inner radius of the ring_2D_kernal used for the convolution.
-      outer : int
-          The outer radius of the ring_2D_kernal used for the convolution.
-      sat_flag : int
-          The integer value of the saturation flag.
-      jump_flag : int
-          The integer value of the jump flag
-      ellipse_expand : float
-          The relative increase in the size of the fitted ellipse to be
-          applied to the shower.
-      num_grps_masked : int
-          The number of groups after the detected shower to be flagged as jump.
-      max_extended_radius : int
-          The upper limit for the extension of saturation and jump
-      minimum_sigclip_groups : int
-          The minimum number of groups to use sigma clipping.
-      max_shower_amplitude : float
-          The maximum amplitude of shower artifacts to correct in DN/group
-
 
     Returns
     -------
     gdq : int, 4D array
-      updated group dq array.
-    number_ellipse : int
-    Total number of showers detected.
+        updated group dq array.
 
+    number_ellipse : int
+        Total number of showers detected.
     """
     log.info("Flagging Showers")
-    refpix_flag = dqflags["REFERENCE_PIXEL"]
+    refpix_flag = jump_data.fl_ref
+
     gdq = ingdq.copy()
     data = indata.copy()
-    nints = data.shape[0]
-    ngrps = data.shape[1]
-    num_grps_donotuse = 0
+    nints, ngrps, nrows, ncols = data.shape
+
+    num_grps_donotuse = count_dnu_groups(gdq, jump_data)
 
-    for integ in range(nints):
-        for grp in range(ngrps):
-            if np.all(np.bitwise_and(gdq[integ, grp, :, :], donotuse_flag)):
-                num_grps_donotuse += 1
     total_diffs = nints * (ngrps - 1) - num_grps_donotuse
     if total_diffs < min_diffs_for_shower:
         log.warning("Not enough differences for shower detections")
         return ingdq, 0
-    read_noise_2 = readnoise_2d**2
-    jump_dnu_flag = jump_flag + donotuse_flag
-    sat_dnu_flag = sat_flag + donotuse_flag
-    data[gdq == jump_dnu_flag] = np.nan
-    data[gdq == sat_dnu_flag] = np.nan
-    data[gdq == sat_flag] = np.nan
-    data[gdq == jump_flag] = np.nan
-    data[gdq == donotuse_flag] = np.nan
+
+    data = nan_invalid_data(data, gdq, jump_data)
+
     refy, refx = np.where(pdq == refpix_flag)
-    gdq[:, :, refy, refx] = donotuse_flag
+    gdq[:, :, refy, refx] = jump_data.fl_dnu
     first_diffs = np.diff(data, axis=1)
 
     all_ellipses = []
 
     first_diffs_masked = np.ma.masked_array(first_diffs, mask=np.isnan(first_diffs))
```

Comment:
masked_array not needed since every usage below uses nanmedian or otherwise ignores nan when collapsing along any axis.  masked_array slows things and adds memory and overhead.

Reply:
The `masked_array` is needed to return the proper class type to be used in the computation of `ratio`.  The `ratio` instance needs the `filled` method, which isn't present in the `first_diffs` instance.  Not using `masked_array` here results in a failure here:

https://github.com/spacetelescope/stcal/blob/5759c19a0ce7cfe66d6230f77c078c84ab516875/src/stcal/jump/jump.py#L1074

Reply:
Strictly true, but since the mask is on nan pixels anyway, the filled method is also redundant and unneeded.  We can defer all of this to performance improvements later.  There is a lot of redundant and/or unneeded work being done in the base branch.